src/container/srv_target.c

```diff
@@ -1142,146 +1148,285 @@ ds_cont_tgt_epoch_discard_aggregator(crt_rpc_t *source, crt_rpc_t *result,
+
+		bufsize = args->snap_count * sizeof(*args->snapshots);
+		D_REALLOC(buf, cont->sc_snapshots, bufsize);
+		if (buf == NULL) {
+			rc = -DER_NOMEM;
+			goto out_cont;
+		}
+		memcpy(buf, args->snapshots, bufsize);
+		cont->sc_snapshots = buf;
 	}
+	cont->sc_snapshots_nr = args->snap_count;
+	cont->sc_aggregation_max = DAOS_EPOCH_MAX;
+	cont->sc_snapshots_ts = crt_hlc_get();
+out_cont:
+	ds_cont_child_put(cont);
+	return rc;
+}
 
-	rc = vos_cont_open(pool_child->spc_hdl, in->tai_cont_uuid,
-			   &cont_hdl);
-	if (rc != 0) {
-		D_ERROR(DF_CONT": Open container failed: %d\n",
-			DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid), rc);
-		goto pool_child;
+int
+ds_cont_tgt_snapshots_update(uuid_t pool_uuid, uuid_t cont_uuid,
+			     uint64_t *snapshots, int snap_count)
+{
+	struct cont_snap_args	 args;
+
+	uuid_copy(args.pool_uuid, pool_uuid);
+	uuid_copy(args.cont_uuid, cont_uuid);
+	args.snap_count = snap_count;
+	args.snapshots = snapshots;
+	D_DEBUG(DB_TRACE, DF_UUID": refreshing snapshots %d\n",
+		DP_UUID(cont_uuid), snap_count);
+	return dss_thread_collective(cont_snap_update_one, &args, 0);
+}
+
+void
+cont_snapshots_refresh_ult(void *data)
+{
+	struct cont_snap_args	*args = data;
+	struct ds_pool		*pool;
+	int			 rc;
+
+	pool = ds_pool_lookup(args->pool_uuid);
+	if (pool == NULL) {
+		rc = -DER_NO_HDL;
+		goto out;
 	}
+	rc = cont_iv_snapshots_refresh(pool->sp_iv_ns, args->cont_uuid);
+	ds_pool_put(pool);
+out:
+	if (rc != 0)
+		D_WARN(DF_UUID": failed to refresh snapshots IV: rc %d;"
+			" Aggregation may not work correctly\n",
+			DP_UUID(args->cont_uuid), rc);
+	D_FREE(args);
+}
 
-	rc = vos_cont_query(cont_hdl, &cont_info);
-	if (rc != 0) {
-		D_ERROR(DF_CONT": Query container failed: %d\n",
-			DP_CONT(in->tai_pool_uuid, in->tai_cont_uuid), rc);
-		goto cont_close;
+int
+ds_cont_tgt_snapshots_refresh(uuid_t pool_uuid, uuid_t cont_uuid)
+{
+	struct cont_snap_args	*args;
+	int			 rc;
+
+	D_ALLOC(args, sizeof(*args));
+	if (args == NULL)
+		return -DER_NOMEM;
+	uuid_copy(args->pool_uuid, pool_uuid);
+	uuid_copy(args->cont_uuid, cont_uuid);
+	rc = dss_ult_create(cont_snapshots_refresh_ult, args,
+			    DSS_ULT_POOL_SRV, 0, 0, NULL);
+	if (rc != 0)
+		D_FREE(args);
+	return rc;
+}
+
+int
+cont_child_aggregate(struct ds_cont_child *cont)
+{
+	daos_epoch_t		 epoch_max;
+	daos_epoch_range_t	 epoch_range;
+	vos_cont_info_t		 cinfo;
+	int			 i;
+	int			 rc;
+
+	rc = vos_cont_query(cont->sc_hdl, &cinfo);
+	if (rc)
+		return rc;
+	/*
+	 * Upper bound for aggregation is the smallest of:
+	 * 1. Limit set when snapshot is in progress
+	 * 2. Lowest active (prepared) DTX epoch
+	 * 3. Highest committed DTX epoch
+	 */
+	epoch_max = MIN(cont->sc_aggregation_max,
+			MIN(cinfo.ci_lpe, cinfo.ci_hce));
+
+	D_ASSERTF(cinfo.ci_hae <= epoch_max,
+		  "Highest aggregated %lu; Max %lu",
+		  cinfo.ci_hae, epoch_max);
+	/*
+	 * Start from the beginning if snapshots were refreshed since
+	 * the last aggregation in order to handle deleted snapshots
+	 */
```

Comment:
What if new snapshot created? Will we start from beginning?

Reply:
What do you mean by that... aggregation would always go from highest aggregated epoch => highest stable epoch; only time it has to go back from the start is on snapshot deletion but I haven't figured out that case yet.

Reply:
I mean I'm wondering if the 'sc_snapshot_ts' will also be updated and causing aggregate from 0 when creating snapshot.
BTW,  in my opinion, aggregation should always go from the nearest snapshot lower than 'highest aggregated epoch' to 'highest stable epoch'. For snapshot deletion, only the epr of [snapshot_prior_deleted, snapshot_after_deleted] needs be aggregated.